crypto/crypto/src/macros/src/lib.rs

```diff
@@ -36,3 +111,327 @@ pub fn silent_debug(source: TokenStream) -> TokenStream {
+                    #to_bytes_arms
+                }
+            }
+        }
+    });
+    try_from.into()
+}
+
+#[proc_macro_derive(PublicKey, attributes(PrivateKeyType))]
+pub fn derive_enum_publickey(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+
+    let name = &ast.ident;
+    let private_key_type = get_type_from_attrs(&ast.attrs, "PrivateKeyType");
+    match ast.data {
+        Data::Enum(ref variants) => impl_enum_publickey(name, private_key_type, variants),
+        Data::Struct(_) | Data::Union(_) => {
+            panic!("#[derive(PublicKey)] is only defined for enums")
+        }
+    }
+}
+
+fn get_type_from_attrs(attrs: &[syn::Attribute], attr_name: &str) -> String {
+    let mut associated_key_type = None;
+    for attr in attrs {
+        if let syn::Meta::NameValue(val) = attr.parse_meta().unwrap() {
+            if val
+                .path
+                .get_ident()
+                .expect(&format!("Could not parse {} attribute", attr_name)[..])
+                == attr_name
+            {
+                if let syn::Lit::Str(lit) = &val.lit {
+                    associated_key_type = Some(lit.value());
+                }
+            }
+        }
+    }
+
+    associated_key_type.expect(&format!("Could not find {} attribute", attr_name)[..])
+}
+
+fn impl_enum_publickey(name: &Ident, private_key_type: String, variants: &DataEnum) -> TokenStream {
+    let pkt = syn::parse_str::<syn::Type>(&private_key_type).expect("Misformatted PrivateKeyType");
+    let mut from_match_arms = quote! {};
+    for variant in variants.variants.iter() {
+        let variant_ident = &variant.ident;
+
+        from_match_arms.extend(quote! {
+            #pkt::#variant_ident(key) => #name::#variant_ident(key.into()),
+        });
+    }
+    let mut res = quote! {
+        impl From<&#pkt> for #name {
+            fn from(public_key: &#pkt) -> Self {
+                match public_key {
+                    #from_match_arms
+                }
+            }
+        }
+    };
+    res.extend(quote! {
+        impl crypto::PublicKey for #name {
+            type PrivateKeyMaterial = #pkt;
+        }
+    });
+    res.into()
+}
+
+#[proc_macro_derive(PrivateKey, attributes(PublicKeyType))]
+pub fn derive_enum_privatekey(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+
+    let name = &ast.ident;
+    let public_key_type = get_type_from_attrs(&ast.attrs, "PublicKeyType");
+    match ast.data {
+        Data::Enum(ref variants) => impl_enum_privatekey(name, public_key_type, variants),
+        Data::Struct(_) | Data::Union(_) => {
+            panic!("#[derive(PrivateKey)] is only defined for enums")
+        }
+    }
+}
+
+fn impl_enum_privatekey(
+    name: &Ident,
+    public_key_type: String,
+    _variants: &DataEnum,
+) -> TokenStream {
+    let pkt = syn::parse_str::<syn::Type>(&public_key_type).expect("Misformatted PrivateKeyType");
```

Comment:
Same as above -- it would be better to parse from a LitStr instead of raw String so that parse errors can appear in the correct place.